File: client/web/src/features/auth/api/auth-api.ts
import { apiClient } from "@/lib/api-client";
import type {
  LoginChallengeResponse,
  LoginCompleteResponse,
  RegisterRequest,
  RegisterResponse,
  BeginRegistrationResponse,
  PasswordChangeResponse,
  DeviceDto,
  Envelope,
  KdfConfig,
  PasswordVerifierDto,
} from "@/types/api-types";

export function requestLoginChallenge(
  email: string,
  device: DeviceDto,
): Promise<LoginChallengeResponse> {
  return apiClient.post<LoginChallengeResponse>("/auth/login", {
    email,
    device,
  });
}

export function completeLogin(
  loginSessionId: string,
  clientProof: string,
): Promise<LoginCompleteResponse> {
  return apiClient.post<LoginCompleteResponse>("/auth/login/complete", {
    loginSessionId,
    clientProof,
  });
}

export function registerBegin(
  email: string,
): Promise<BeginRegistrationResponse> {
  return apiClient.post<BeginRegistrationResponse>("/auth/register/begin", { email });
}

export function verifyEmail(
  registrationSessionId: string,
  verificationCode: string,
): Promise<{ verified: boolean }> {
  return apiClient.post<{ verified: boolean }>("/auth/register/verify-email", {
    registrationSessionId,
    verificationCode,
  });
}

export function register(
  request: RegisterRequest,
): Promise<RegisterResponse> {
  return apiClient.post<RegisterResponse>("/auth/register", request);
}

// --- Password Reset ---

export function beginPasswordReset(
  email: string,
  device?: DeviceDto,
): Promise<{ resetSessionId: string; codeSent: boolean; verificationCode?: string }> {
  return apiClient.post("/auth/password/reset/begin", { email, device });
}

export function verifyPasswordResetCode(
  resetSessionId: string,
  verificationCode: string,
): Promise<{ verified: boolean; kdf?: KdfConfig; deviceEnvelope?: Envelope }> {
  return apiClient.post("/auth/password/reset/verify", { resetSessionId, verificationCode });
}

export function completePasswordReset(
  resetSessionId: string,
  newVerifier: PasswordVerifierDto,
  newPasswordEnvelope: Envelope,
): Promise<LoginCompleteResponse> {
  return apiClient.post("/auth/password/reset/complete", { resetSessionId, newVerifier, newPasswordEnvelope });
}

// --- Password Change ---

export function changePassword(
  currentProof: { loginSessionId: string; clientProof: string },
  newVerifier: PasswordVerifierDto,
  newPasswordEnvelope: Envelope,
): Promise<PasswordChangeResponse> {
  return apiClient.post<PasswordChangeResponse>("/account/password/change", {
    currentProof,
    newVerifier,
    newPasswordEnvelope,
  });
}

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/AuthApi.kt
package io.github.pstanar.pstotp.core.api

import io.github.pstanar.pstotp.core.model.api.BeginRegistrationResponse
import io.github.pstanar.pstotp.core.model.api.DeviceDto
import io.github.pstanar.pstotp.core.model.api.Envelope
import io.github.pstanar.pstotp.core.model.api.LoginChallengeResponse
import io.github.pstanar.pstotp.core.model.api.LoginCompleteResponse
import io.github.pstanar.pstotp.core.model.api.PasswordVerifierDto
import io.github.pstanar.pstotp.core.model.api.RecoveryDto
import io.github.pstanar.pstotp.core.model.api.RegisterResponse
import org.json.JSONObject

/** Authentication API endpoints (public — no Bearer token). */
class AuthApi(private val client: ApiClient) {

    suspend fun requestLoginChallenge(email: String, device: DeviceDto): LoginChallengeResponse {
        val body = JSONObject().apply {
            put("email", email)
            put("device", device.toJson())
        }
        return client.postPublic("/auth/login", body) { LoginChallengeResponse.fromJson(it) }
    }

    suspend fun completeLogin(loginSessionId: String, clientProof: String): LoginCompleteResponse {
        val body = JSONObject().apply {
            put("loginSessionId", loginSessionId)
            put("clientProof", clientProof)
        }
        return client.postPublic("/auth/login/complete", body) { LoginCompleteResponse.fromJson(it) }
    }

    suspend fun registerBegin(email: String): BeginRegistrationResponse {
        val body = JSONObject().put("email", email)
        return client.postPublic("/auth/register/begin", body) { BeginRegistrationResponse.fromJson(it) }
    }

    suspend fun verifyEmail(registrationSessionId: String, verificationCode: String): Boolean {
        val body = JSONObject().apply {
            put("registrationSessionId", registrationSessionId)
            put("verificationCode", verificationCode)
        }
        return client.postPublic("/auth/register/verify-email", body) { it.getBoolean("verified") }
    }

    suspend fun register(
        registrationSessionId: String?,
        email: String,
        passwordVerifier: PasswordVerifierDto,
        passwordEnvelope: Envelope,
        device: DeviceDto,
        deviceEnvelope: Envelope,
        recovery: RecoveryDto,
    ): RegisterResponse {
        val body = JSONObject().apply {
            if (registrationSessionId != null) put("registrationSessionId", registrationSessionId)
            put("email", email)
            put("passwordVerifier", passwordVerifier.toJson())
            put("passwordEnvelope", passwordEnvelope.toJson())
            put("device", device.toJson())
            put("deviceEnvelope", deviceEnvelope.toJson())
            put("recovery", recovery.toJson())
        }
        return client.postPublic("/auth/register", body) { RegisterResponse.fromJson(it) }
    }

    suspend fun changePassword(
        currentProof: JSONObject,
        newVerifier: JSONObject,
        newPasswordEnvelope: JSONObject,
    ): JSONObject {
        val body = JSONObject().apply {
            put("currentProof", currentProof)
            put("newVerifier", newVerifier)
            put("newPasswordEnvelope", newPasswordEnvelope)
        }
        return client.post("/account/password/change", body) { it }
    }

    suspend fun logout() {
        val body = JSONObject().apply {
            client.refreshToken?.let { put("refreshToken", it) }
        }
        try {
            client.postPublic("/auth/logout", body) { }
        } catch (_: Exception) {
            // Best-effort — server may be unreachable
        }
    }
}
